cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+used to generate Rust source.
+
+The decision trie is, as the name implies, a kind of decision tree, in
+the sense that we start at the root and move down the tree based on
+the result of match operations (each feeding one "decision").
+
+It is a "trie" (which is a kind of tree) because at each level, its
+edges are labeled with match operations; a trie is a tree where one
+input character from an alphabet is used to index children at each
+level.
+
+Each node in the tree is either an internal decision node, or a leaf
+"expression" node (which we reach once we have a successful rule
+match). The "execution semantics" of the trie are
+backtracking-based. We attempt to find some path down the tree through
+edges whose match ops run successfully; when we do this to reach a
+leaf, we have the values generated by all of the match ops, and we can
+execute the sequence of "expression instructions" in the leaf. Each
+rule's left-hand side becomes a series of edges (merged into the
+existing tree as we process rules) and each rule's right-hand side
+becomes one leaf node with expression instructions.
+
+At any point, if a match op does not succeed, we try the next out-edge
+in sequence. If we have tried all out-edges from a decision node and
+none were successful, then we backtrack one level further. Thus, we
+simply perform an in-order tree traversal and find the first
+successful match.
+
+Though this sounds possibly very inefficient if some decision node has
+a high fan-out, in practice it is not because the edges are often
+known to be *mutually exclusive*. The canonical example of this is
+when an enum-typed value is destructured into different variants by
+various edges; we can use a Rust `match` statement in the generated
+source and have `O(1)` (or close to it) cost for the dispatch at this
+level.
+
+Building the trie is a somewhat subtle procedure because we must be
+mindful of *priorities*. In order to ensure that rule priorities are
+obeyed, we actually label each edge with a *priority range* and a
+match op. Edges must then be considered in priority order. When
+building the trie, we uphold an important *non-overlapping-priority*
+invariant: there must not be two edges with overlapping priority
+ranges and different match ops, or else it would be unclear which to
+run next. Priority ranges are inclusive in a discrete (integer) space,
+so "overlap" is also subtle: we can allow edges with "unit" priority
+ranges (just one priority), and these do not overlap. So, e.g., we can
+allow an edge with priority `[0, 10]` (higher is more prioritized) and
+op `A`, followed by an edge with priority `[0, 0]` and op `B`,
+followed by an edge with priority `[-1, -10]`, `A`, though they must
+occur in that order.
+
+See [this block
+comment](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/isle/isle/src/trie.rs#L15-L166)
+for more information regarding the trie construction algorithm.
+
+## Reference: ISLE Language Grammar
+
```

Comment:
Questions/comments at this point:
- Might be nice to show how this decision trie algorithm handles the more-and-less specific rule examples from the beginning. In practice, what does the prioritization mean for how lowering instructions are selected?
- Maybe add a brief reference that instruction selection is in general NP-complete, so accepting backtracking/some level of possible inefficiency is unavoidable. 
- What should engineers consider when adding new ISEL rules? Is all of Cranelift's lowering handled by ISEL (I assume no?) and what does the interaction between ISEL and the more traditional instruction selection backend look like? On this point, might be worth clarifying which invariants are enforced by the compiler vs the engineer (as I think I commented in another place, having ISEL or the Rust compiler itself rule out side effects in the LHS seems valuable, as does enforcement of which sublanguage can be on the LHS/RHS).

Reply:
> show how this decision trie algorithm handles the more-and-less specific rule examples from the beginning
 
There's a bit more discussion in the block comment that's referenced now, but as far as the more-specific-first ordering, it's a bit subtle. IMHO the best way may not be to try to get the programmer to predict the current algorithm but just to make it a more explicit and well-defined ordering. I created #3575 for this.

> brief reference that instruction selection is in general NP-complete

Added a note that collapsing the rule patterns to a fast-dispatch `match` is not guaranteed, and that worst-case complexity per rewrite is equivalent to a sequential attempt at each rule in turn. The way that we integrate ISLE into our lowering/isel framework I think takes care of the weighted set cover kind of NP-completeness that one generally has when doing isel over the dataflow graph (we do greedy matching) but the complexity is certainly still something to be aware of.

> What should engineers consider when adding new ISLE rules? [interaction between ISLE and traditional isel, other details]

@fitzgen's doc I think covers this pretty well, hopefully! If not then we can definitely add more pointers/tips as we encounter/come up with them.